### DISCUSSION TITLE
Can a routing guide file (generated by a global router) be loaded into OpenROAD, and then perform parasitic extraction followed by STA?

### DISCUSSION THREAD

**Discussion by liangrj2014** [The-OpenROAD-Project/OpenROAD-flow-scripts#2197] [Q&A] [answered]
From the scripts here: [OpenROAD-flow-scripts/flow/scripts/global_route.tcl at master · The-OpenROAD-Project/OpenROAD-flow-scripts (github.com)](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/master/flow/scripts/global_route.tcl) we can see that the parasitic extraction followed by STA can be conducted at the global routing stage, and the routing guide can be dumped out after global routing. I wonder whether we can load a routing guide file (generated by other global router) into OpenROAD, and then perform parasitic extraction and STA?

**Comment by maliberty** [comment]
I would think so but have never tried it.

**Reply by liangrj2014** [reply]
Thanks! I would give a try :)

**Comment by maliberty** [comment]
Note the guides need to look like routing - wide areas will not work well.

**Reply by liangrj2014** [reply]
Thanks for the suggestion! I ran an experiment as follows:

1) Load a placed design to OpenROAD and then run global_route to generate route guide. No errors or warning are reported. Also, I conduct parasitic estimation ("set_propagated_clock [all_clocks]; 
 estimate_parasitics -global_routing") and report timing. Also no errors or warning are reported;

2) Open a new terminal. Load the placed design to OpenROAD and read the route guide generated by 1). Still no errors or warnings. Then I conduct parasitic estimation. Now I get a few warnings saying "Missing route to pin XXX". Then I report timing to get wns/tns, which are significantly different than the wns/tns in 1).

Do you have any ideas in where goes wrong and how to address the mismatch? Thanks!

**Reply by maliberty** [reply]
The wns/tns will depend on the routes themselves (poor routing affects timing).  For a sanity check you could write out guides and then read the same back in.

The uncovered pin messages happen when a pin straddles multiple gcells and you route to a different one than the router considers the pin to be in.

**Reply by liangrj2014** [reply]
Actually this is exactly what I did: "write out guides and then read the same back in". But I got different timing :(

**Reply by maliberty** [reply]
What design did you use?

**Reply by liangrj2014** [reply]
I used design aes from IWLS 2005 benchmark suits, with ASAP7 library

**Reply by liangrj2014** [reply]
I could share design files if it would help :)

**Reply by liangrj2014** [reply]
Hi! Are there any suggestions on this issue? Thanks in advance!

**Comment by maliberty** [comment]
@eder-matheus would you or @luis201420 take a look

**Reply by eder-matheus** [reply]
@maliberty @liangrj2014 The global router doesn't support parasitics extraction from loaded guide files. Currently, we support guide visualization and the congestion heatmap with the data from loaded guide files.

The issue is that we lost a lot of information about routing connectivity when creating the guides. We remove guides overlapping in the same layer, we don't have guides for vias, and reconstructing these information from the guides might be tricky.

One alternative could be to dump the route segments constructed for each net (i.e., the raw information that the guides are constructed from). It would be easy to read this and fill the necessary structs to perform parasitic extraction, repair antennas, incremental groute, etc.

**Reply by liangrj2014** [reply]
Thanks for the feedback! Which command to dump out and read back the route segments? And what format to store the route segments information? Thanks!

**Reply by maliberty** [reply]
Sorry it sounds like I was mistaken on it being easy.

**Reply by eder-matheus** [reply]
@liangrj2014 Sorry, I wasn't clear. This command doesn't exist yet, but we can implement it. @maliberty Do you think it's reasonable to have such a command to dump the route segments? It wouldn't be hard to implement, in my opinion.

**Reply by liangrj2014** [reply]
It will be really helpful if we can have such commands! We are helping to organize the ISPD2025 contest on performance-driven global routing and looking for an open-source tool that can evaluate PPA of global routing solutions. It will be wonderful if OpenROAD could load a global routing solution, extract RC and analyze timing/power!

**Reply by eder-matheus** [reply]
@liangrj2014 Here are the two commands that you can use to write and read the global route segments and perform parasitic extraction from them. Let me know if you have any other questions.

**Reply by maliberty** [reply]
@eder-matheus 'here' is unclear

**Reply by eder-matheus** [reply, marked as answer]
> @eder-matheus 'here' is unclear

I'm sorry, I forgot to paste the link: https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/src/grt/README.md#write-global-routing-segments

**Reply by liangrj2014** [reply]
Thanks for the great help! Let me give a try soon :)

**Reply by liangrj2014** [reply]
Thank you once again for implementing the new commands! I conducted a brief test by writing out a route segments file and then reading it back in. The process ran smoothly, and I observed consistent timing results before and after the read/write operations :) 

Upon examining the route segments file, I noticed that its format appears similar to the route guide format. Is it possible to provide a detailed description of the route segments format? Specifically, I am interested in understanding the syntax used and the criteria for valid route segments solutions. This information would be invaluable for participants in the ISPD contest to create route segments files compatible to the OpenROAD flow. Thanks in advance!

**Reply by eder-matheus** [reply]
@liangrj2014 The main difference between the route guide and route segments format is what is described.

In the route guides, each line represents a rectangle in a specific routing layer: `(lower_left_x, lower_left_y, upper_right_x, upper_right_y, routing_layer)`.

In the route segments, each line describes a vertical or horizontal line, or a via, going from one point to another in a given routing layer: `(x1, y1, routing_layer1, x2, y2, routing_layer2)`.

Some examples:
1) `18000 154800 met4 18000 205200 met4`: this segment represents a vertical line in the routing layer met4.
2) `18000 154800 met3 133200 154800 met3`: this segment represents a horizontal line in the routing layer met3.
3) `18000 205200 met3 18000 205200 met4`: this segment represents a via going from met3 to met4.
4) `133200 154800 met3 133200 154800 met2`: this segment represents a via going from met3 to met2.
Notice that the via segments can go from lower layers to upper layers (like in example 3) or from upper layers to lower layers (like in example 4).

To perform parasitic extraction from the route generated in other global route tools, you would also need to give OpenROAD's global router the pin positions of each net relative to the global routing grid of the external tools. This requires a bit of background on how our global router works, so I'll explain it in a separate comment.

**Reply by eder-matheus** [reply]
Our global router creates a routing grid based on the size of each GCell. The grid looks like this:
<img width="40%" alt="image" src="https://github.com/user-attachments/assets/f6c1d807-8bf4-42a4-b794-6fc9ba8b7afe">

Each square of the grid is a grid point. The specific coordinates of each point are the center of the squares. When the global router parses the nets of the design, each pin of the net is "translated" to a position in the routing grid. In the figure below, we see that the highlighted pin `Y` have its position in the global routing grid in the yellow dot, the center of the square that contains the larger part of the pin geometry:
<img width="30%" alt="Screenshot 2024-07-31 at 13 07 17" src="https://github.com/user-attachments/assets/1b4d6963-4c04-4daa-a2c9-ce0200caa076">

This is important because when extracting parasitics from the global route segments, the tool will make connections between the segments and these positions in the routing grid. In other words, the segments must be connected to the position that represents the pins in the routing grid.

**Reply by eder-matheus** [reply]
Different global routers might assign the pins to the routing grid in different ways, but for extracting parasitics, we only need the final pin position on the grid. So I believe that the input for OpenROAD, being the route segments and the pin coordinates in the routing grid, should be enough to perform parasitics extraction with no errors.

**Reply by liangrj2014** [reply]
Thanks @eder-matheus for your explanation! 

Regarding "the pin positions of each net relative to the global routing grid", would OpenROAD infer this information from the placed DEF file, provided that the external tools use the same GCell definition as those defined in the placed DEF file? 

Additionally, I am curious to know if OpenROAD verifies the legality of the loaded route segment files. For instance, does it check whether the route segments for a net successfully connect all the pins of the net?

**Reply by eder-matheus** [reply]
We have this function that calculates the pin position on grid: https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/src/grt/src/GlobalRouter.cpp#L767-L805
But as I mentioned, it's possible that other tools use other approaches to compute that, making it differ from what OpenROAD calculates internally.

About if it checks for the legality of the loaded route segments, not directly. If you have an invalid segments file as input, you might face warnings during the parasitics extraction, and the results may not be correct. But we don't have an explicit checker for this format yet.

**Reply by liangrj2014** [reply]
Got it! Really appreciate for the detailed explanations and all the helps!!!!

**Reply by eder-matheus** [reply]
@liangrj2014 FYI, we recently added support for checking input route segments. The PR is this one: https://github.com/The-OpenROAD-Project/OpenROAD/pull/5518.

Now, the tool will error if you load route segments that are not connected or do not cover a pin in the design.

**Comment by b224hisl** [comment]
I would like to ask whether OpenROAD supports manually setting up the route guide (not just a useless output). If I want to manually route some nets, how should I proceed?

**Reply by maliberty** [reply]
@eder-matheus please respond.

**Reply by eder-matheus** [reply]
@b224hisl I'm not sure if I understand the question. Do you want to manually create the route guides of a set of nets? I don't think we have such capability yet. You could try to manually write the guide file, where the units are in DBU, but we don't have internal commands that would help with that.

If you could explain your use case with more details, we can try to implement such capability.

**Reply by b224hisl** [reply]
Based on my understanding, the route guide file is merely an output of the global routing process. Even if I were to manually create this route guide, it seems that it cannot be read during the routing stage and used to guide the routing within the specified regions. If it is indeed possible to do so, could you kindly let me know how this can be achieved? Thank you for your assistance!

**Reply by eder-matheus** [reply]
You can use the [read_guides](https://github.com/The-OpenROAD-Project/OpenROAD/tree/master/src/grt#read-global-routing-guides) command. The input guides will then be used during the detail routing tool.
